modules/clients/src/test/java/org/apache/ignite/jdbc/JdbcAuthorizationTest.java

```diff
@@ -0,0 +1,440 @@
+    }
+
+    /**
+     * Inserts random key into the test table.
+     *
+     * @return Inserted key.
+     */
+    private int insertKey() throws Exception {
+        int key = UUID.randomUUID().hashCode();
+
+        execute(of("INSERT INTO " + TEST_DML_TABLE + "(_key, _val) VALUES (" + key + ", 0);"), CACHE_PUT_USER);
+
+        return key;
+    }
+
+    /**
+     * Creates table with the specified name.
+     *
+     * @param name Name of the table to be created.
+     */
+    private void createTable(String name) throws Exception {
+        execute(of("CREATE TABLE " + name + "(id LONG PRIMARY KEY, str_col varchar, long_col LONG)" +
+                " WITH \"TEMPLATE=REPLICATED\";"),
+            CACHE_CREATE_SYS_PERM_USER);
+    }
+
+    /**
+     * Asserts that authorization of SQL query on behalf of the user with the specified login.
+     *
+     * @param stmt Provider of the SQL statement.
+     * @param login Login of the user.
+     */
+    private void assertAuthorizationFailed(StatementProvider stmt, String login) {
+        assertThrowsAnyCause(
+            log,
+            () -> {
+                execute(stmt, login);
+
+                return null;
+            },
+            SQLException.class,
+            "Authorization failed");
+    }
+
+    /**
+     * Executes an SQL statement on behalf of a user with the specified login.
+     *
+     * @param stmt Provider of the SQL statement.
+     * @param login Login of the user.
+     */
+    private void execute(StatementProvider stmt, String login) throws Exception {
+        try (Connection conn = getConnection(JDBC_URL_PREFIX + Config.SERVER, login, "")) {
+            stmt.statement(conn).execute();
+        }
+    }
+
+    /**
+     * @param idx Index of the node.
+     * @param clients Security data of the clients.
+     */
+    private IgniteEx startSecurityGrid(int idx, TestSecurityData... clients) throws Exception {
+        String login = getTestIgniteInstanceName(idx);
+
+        return startGrid(getConfiguration(
+            login,
+            new TestSecurityPluginProvider(
+                login,
+                "",
+                systemPermissions(CACHE_CREATE, JOIN_AS_SERVER),
+                null,
+                false,
+                clients
+            ))
+        );
+    }
+
+    /**
+     * @return {@link SecurityPermissionSet} containing specified system permissions.
+     */
+    private SecurityPermissionSet systemPermissions(SecurityPermission... perms) {
+        return create().defaultAllowAll(false).appendSystemPermissions(perms).build();
+    }
+
+    /**
+     * @return {@link SecurityPermissionSet} containing specified cache permissions.
+     */
+    private SecurityPermissionSet cachePermissions(String cache, SecurityPermission... perms) {
+        return create().defaultAllowAll(false).appendCachePermissions(cache, perms).build();
+    }
+
+    /**
+     * Provider of SQL statement.
+     */
+    static interface StatementProvider {
```

Comment:
StatementProvided implemented only twice, in the `of()` method and in the `testBulkInsert` method (which is not correct). I think this interface should be removed and `String sql` should be passed to `execute()` method instead.

Reply:
Thanks. Done.